feat(filesystem): customizable timestamp formats

MastarCheeze opened the pull request:
I've added a `format` field in the configs for the components `last_modified` and `created` so that users can change the format string used to display timestamps in the file explorer. The field also accepts a function that returns a string so that features like relative timestamps can be implemented. The change is also applied to the file details popup window.

Here's an example setup:

```lua
require("neo-tree").setup({
  default_component_configs = {
    last_modified = {
      enabled = true,
      format = function(seconds)
        local now = os.time()
        local diff = now - seconds

        local function pluralise(value, unit)
          return value .. " " .. unit .. (value == 1 and "" or "s") .. " ago"
        end

        if diff < 60 then
          return "Just now"
        elseif diff < 3600 then
          local minutes = math.floor(diff / 60)
          return pluralise(minutes, "minute")
        elseif diff < 86400 then
          local hours = math.floor(diff / 3600)
          return pluralise(hours, "hour")
        elseif diff < 86400 * 30 then
          local days = math.floor(diff / 86400)
          return pluralise(days, "day")
        elseif diff < 86400 * 365 then
          local months = math.floor(diff / (86400 * 30))
          return pluralise(months, "month")
        else
          local years = math.floor(diff / (86400 * 365))
          return pluralise(years, "year")
        end
      end,
    },
    created = {
      enabled = true,
      format = "%e %b %y at %I:%M %p",
    },
  },
})
```

![File explorer](https://github.com/user-attachments/assets/8c35e1fa-f2a1-4584-af69-3470afd5eaae)
![File details popup window](https://github.com/user-attachments/assets/cce0eda8-c7d5-4d78-a57e-408ef561f00b)

The columns are offset in the image because the the timestamp is longer than the column heading. I'll be opening a separate pull request that deals with that.

MastarCheeze commented:
I just found out that there is already an existing issue (#1505) and a PR (#1506) for this exact topic.

pynappo commented on lua/neo-tree/defaults.lua:
```diff
@@ -257,10 +257,14 @@ local config = {
     last_modified = {
       enabled = true,
       required_width = 88, -- min width of window required to show this column
+      format = "%Y-%m-%d %I:%M %p", -- format string for timestamp (see ISO C function `strftime`, `format` parameter)
+                                    -- or use a function that takes in the date in seconds and returns a string
     },
     created = {
       enabled = false,
       required_width = 120, -- min width of window required to show this column
+      format = "%Y-%m-%d %I:%M %p", -- format string for timestamp (see ISO C function `strftime`, `format` parameter)
+                                    -- or use a function that takes in the date in seconds and returns a string
```
There's builtin help for lua's `os.date()` (which also leads to vimscript's `strftime()`) so that might be a bit more convenient to refer to:
```suggestion
      format = "%Y-%m-%d %I:%M %p", -- format string for timestamp (see `:h os.date()`)
                                    -- or use a function that takes in the date in seconds and returns a string
    },
    created = {
      enabled = false,
      required_width = 120, -- min width of window required to show this column
      format = "%Y-%m-%d %I:%M %p", -- format string for timestamp (see `:h os.date()`)
                                    -- or use a function that takes in the date in seconds and returns a string
```

MastarCheeze replied:
Alright, I'll update the documentation.

pynappo commented on lua/neo-tree/sources/common/commands.lua:
```diff
@@ -530,9 +530,19 @@ M.show_file_details = function(state)
     table.insert(right, utils.human_size(stat.size))
     table.insert(left, "Created")
-    table.insert(right, os.date("%Y-%m-%d %I:%M %p", stat.birthtime.sec))
+    local created_format = require("neo-tree").config.default_component_configs.created.format
```
hm, maybe it would be useful if this (and `modified_format`) were to be configurable in the mapping level as well? could be something like:

```suggestion
    local created_format = state.config.created_format or require("neo-tree").config.default_component_configs.created.format
```

Then the user could optionally specify:

```lua
mappings = {
  ['I'] = {
    'show_file_details',
    config = {
      created_format = '...',
      modified_format = '...'
    }
  }
}
```

so the details popup can time differently than the file tree.

It would require more documentation in `defaults.lua`, the vim help doc, and maybe the README though. LMK your thoughts on it.

MastarCheeze replied:
Sounds good!

I think the mappings config should be separate from the default component config to avoid confusion. Changes to the mappings config should only affect the file details popup, and changes to the component config should only affect the info column in the file explorer.

I'll work on an implementation and add in the necessary documentation soon!

MastarCheeze commented:
I've added the changes you suggested @pynappo.

On second thought, I think its better to use `default_component_configs.created.format` and `default_component_configs.modified.format` as fallback values for the mapping level configs, as you suggested, to reduce the amount of configuration the user has to do. I've implemented this in my commit, along with documentation.

Thanks for the suggestions!

MastarCheeze commented:
I'm thinking of adding relative timestamps as a builtin feature that can be enabled by specifying:

```lua
default_component_configs = {
  last_modified = {
    relative = true
  }
}
```

Or perhaps like this:

```lua
default_component_configs = {
  last_modified = {
    format = "relative"
  }
}
```

Or even just include a commented-out implementation of a relative timestamp function in `README.md`.

Any opinions about this? Should this be in its own PR?

pynappo commented on lua/neo-tree/sources/common/commands.lua:
```diff
@@ -530,9 +530,21 @@ M.show_file_details = function(state)
     table.insert(right, utils.human_size(stat.size))
     table.insert(left, "Created")
-    table.insert(right, os.date("%Y-%m-%d %I:%M %p", stat.birthtime.sec))
+    local created_format = state.config.created_format
+      or require("neo-tree").config.default_component_configs.created.format
+    if type(created_format) == "function" then
+      table.insert(right, created_format(stat.birthtime.sec))
+    else
+      table.insert(right, os.date(created_format, stat.birthtime.sec))
+    end
```
I think you could pull this logic out into a new file like `sources/common/lib/format.lua` just for code reuse purposes:

```lua
-- sources/common/lib/format.lua
local M = {}
M.date = function(format, seconds)
  -- add logic for special case for format = "relative"
  return type(format) == "function" and format(seconds) or os.date(format, seconds)
end
return M 
```

```suggestion
    local date = require('neotree.sources.common.lib.format').date
    local created_format = state.config.created_format
      or require("neo-tree").config.default_component_configs.created.format
    table.insert(right, date(created_format, stat.birthtime.sec))
    local modified_format = state.config.modified_format
      or require("neo-tree").config.default_component_configs.modified.format
    table.insert(right, date(modified_format, stat.mtime.sec))
```

MastarCheeze replied:
It would be cleaner to implement relative timestamps in the utils module as you suggested, then add a commented-out setting in `defaults.lua`.

```lua
-- lua/neo-tree/utils/init.lua
M.relative_date = function(seconds)
  -- Implementation
end
```

```lua
-- lua/neo-tree/defaults.lua
...
last_modified = {
  ...
  --format = require("neo-tree.utils").relative_date
},
...
```

What do you think? I'll implement this along with this PR.

pynappo replied:
sounds good to me!

MastarCheeze commented:
All done! I've added the suggested changes.

MastarCheeze commented:
Any updates on this @pynappo?

pynappo commented:
Apologies, had to think about the best way to reduce the amount of configuration needed. I've decided to remove the fallback of file_details's timestamps on the component configs because I think it should always default to giving a detailed absolute time. I also added `format = "relative"` as shorthand.
